File: backend/tools/semantic_fs.py
from __future__ import annotations
from typing import Optional, Any, Callable
from mixedbread import Mixedbread
from datetime import datetime
import json


class SemanticFS:
    """
    A semantic filesystem backed by mixedbread.

    Paths like:
      /todos/buy-groceries.md
      /memories/user.md
      /tools/add_todo.md

    Supports CRUD operations and semantic search across all or filtered paths.
    """

    def __init__(self, api_key: str, store_name: str = "bernd"):
        self.mixedbread = Mixedbread(api_key=api_key)
        self.store_name = store_name
        self._setup_store()

        # For toolbox functionality
        self._tool_handlers: dict[str, Callable] = {}
        self._tool_schemas: dict[str, dict] = {}

    def _setup_store(self):
        """Create store if it doesn't exist."""
        store = self.mixedbread.stores.list(q=self.store_name)
        if not store.data:
            self.mixedbread.stores.create(
                name=self.store_name,
                description="Semantic filesystem for Bernd",
                config={
                    "contextualization": {
                        "with_metadata": [
                            "type",
                            "created_at",
                            "updated_at",
                            "priority",
                            "status",
                            "due_date",
                            "tags",
                        ]
                    }
                },
            )

    def _path_to_id(self, path: str) -> str:
        """Convert path to file identifier."""
        path = path.strip()
        if path.startswith("/"):
            path = path[1:]
        return path.replace("/", "__")

    def _id_to_path(self, file_id: str) -> str:
        """Convert file identifier back to path."""
        return "/" + file_id.replace("__", "/")

    # Core filesystem operations

    def write(self, path: str, content: str, metadata: Optional[dict] = None) -> dict:
        """Write content to a path."""
        file_id = self._path_to_id(path)
        now = datetime.now().isoformat()

        # Preserve created_at if file exists
        created_at = now
        try:
            existing = self.mixedbread.stores.files.get(
                file_identifier=file_id,
                store_identifier=self.store_name,
            )
            created_at = existing.metadata.get("created_at", now)
            self.mixedbread.stores.files.delete(
                file_identifier=file_id,
                store_identifier=self.store_name,
            )
        except Exception:
            pass

        meta = metadata or {}
        meta["path"] = path  # Store path for search retrieval
        meta["created_at"] = created_at
        meta["updated_at"] = now

        filename = path.split("/")[-1]
        mime_type = "text/plain" if filename.endswith(".json") else "text/markdown"

        self.mixedbread.stores.files.upload(
            store_identifier=self.store_name,
            file=(filename, content, mime_type),
            metadata=meta,
            external_id=file_id,
            overwrite=True,
        )
        return {"status": "ok", "path": path}

    def read(self, path: str) -> dict:
        """Read content from a path."""
        file_id = self._path_to_id(path)

        try:
            # Get file metadata
            resp = self.mixedbread.stores.files.retrieve(
                file_identifier=file_id,
                store_identifier=self.store_name,
            )
            # Download actual content
            content_resp = self.mixedbread.files.content(file_id=resp.id)
            content = content_resp.read().decode("utf-8")

            return {
                "path": path,
                "content": content,
                "metadata": resp.metadata if hasattr(resp, "metadata") else {},
            }
        except Exception as e:
            print(f"[SemanticFS] Error reading {path}: {e}")
            return {"error": f"Not found: {path}"}

    def delete(self, path: str) -> dict:
        """Delete a file at path."""
        file_id = self._path_to_id(path)
        try:
            self.mixedbread.stores.files.delete(
                file_identifier=file_id,
                store_identifier=self.store_name,
            )
            return {"status": "deleted", "path": path}
        except Exception:
            return {"error": f"Not found: {path}"}

    def list(self, prefix: str = "/", limit: int = 100) -> list[dict]:
        """List files under a path prefix."""
        prefix_id = self._path_to_id(prefix) if prefix != "/" else ""
        files = []
        cursor = None

        while len(files) < limit:
            resp = self.mixedbread.stores.files.list(
                store_identifier=self.store_name,
                limit=min(100, limit - len(files)),
                after=cursor,
            )
            for item in resp.data:
                file_id = item.external_id or ""
                if prefix == "/" or file_id.startswith(prefix_id):
                    files.append(
                        {
                            "path": self._id_to_path(file_id),
                            "metadata": item.metadata
                            if hasattr(item, "metadata")
                            else {},
                        }
                    )
            if not resp.pagination.has_more:
                break
            cursor = resp.pagination.last_cursor

        return files

    def search(self, query: str, prefix: str = "/", top_k: int = 10) -> list[dict]:
        """Semantic search, optionally filtered by path prefix."""
        # Build metadata filter for prefix
        metadata_filter = None
        if prefix != "/":
            metadata_filter = {
                "key": "path",
                "operator": "starts_with",
                "value": prefix,
            }

        try:
            resp = self.mixedbread.stores.search(
                store_identifiers=[self.store_name],
                query=query,
                top_k=top_k,
                filters=metadata_filter,
            )
        except Exception as e:
            print(f"[SemanticFS] Search error: {e}")
            return []

        results = []
        for item in resp.data:
            metadata = item.metadata if hasattr(item, "metadata") else {}
            path = metadata.get("path", "/" + (item.filename or ""))

            results.append(
                {
                    "path": path,
                    "content": item.text if hasattr(item, "text") else "",
                    "score": item.score,
                    "metadata": metadata,
                }
            )

        return results

    def clear_prefix(self, prefix: str) -> dict:
        """Delete all files under a prefix."""
        files = self.list(prefix=prefix, limit=1000)
        deleted = 0
        for f in files:
            self.delete(f["path"])
            deleted += 1
        return {"status": "cleared", "prefix": prefix, "deleted": deleted}

    def write_binary(self, path: str, data: bytes, mime_type: str, metadata: Optional[dict] = None) -> dict:
        """Write binary data (like images) to a path."""
        file_id = self._path_to_id(path)
        now = datetime.now().isoformat()

        # Delete existing file if present
        try:
            self.mixedbread.stores.files.delete(
                file_identifier=file_id,
                store_identifier=self.store_name,
            )
        except Exception:
            pass

        meta = metadata or {}
        meta["path"] = path
        meta["created_at"] = now
        meta["type"] = "binary"

        filename = path.split("/")[-1]

        self.mixedbread.stores.files.upload(
            store_identifier=self.store_name,
            file=(filename, data, mime_type),
            metadata=meta,
            external_id=file_id,
            overwrite=True,
        )
        return {"status": "ok", "path": path}

    def read_binary(self, path: str) -> dict:
        """Read binary data from a path."""
        file_id = self._path_to_id(path)

        try:
            resp = self.mixedbread.stores.files.retrieve(
                file_identifier=file_id,
                store_identifier=self.store_name,
            )
            content_resp = self.mixedbread.files.content(file_id=resp.id)
            data = content_resp.read()

            return {
                "path": path,
                "data": data,
                "metadata": resp.metadata if hasattr(resp, "metadata") else {},
            }
        except Exception as e:
            print(f"[SemanticFS] Error reading binary {path}: {e}")
            return {"error": f"Not found: {path}"}

    # Toolbox functionality (tools are stored at /tools/*)

    def register_tool(
        self, name: str, description: str, parameters: dict, handler: Callable
    ):
        """Register a tool, storing its schema in the filesystem."""
        self._tool_handlers[name] = handler
        self._tool_schemas[name] = {
            "name": name,
            "description": description,
            "parameters": parameters,
        }

        # Store in filesystem for semantic search
        content = f"Tool: {name}\n\n{description}\n\nParameters:\n{json.dumps(parameters, indent=2)}"
        self.write(f"/tools/{name}.md", content, {"type": "tool"})

    def search_tools(self, query: str, top_k: int = 5) -> list[dict]:
        """Search for tools by intent."""
        results = self.search(query, prefix="/tools", top_k=top_k)
        tools = []
        for r in results:
            name = r["path"].split("/")[-1].replace(".md", "")
            if name in self._tool_schemas:
                tools.append(
                    {
                        "name": name,
                        "description": self._tool_schemas[name]["description"],
                        "score": r["score"],
                    }
                )
        return tools

    def get_tool(self, name: str) -> dict:
        """Get tool schema."""
        if name not in self._tool_schemas:
            return {"error": f"Tool not found: {name}"}
        return self._tool_schemas[name]

    def use_tool(self, name: str, arguments: dict) -> Any:
        """Execute a tool."""
        if name not in self._tool_handlers:
            return {"error": f"Tool not found: {name}"}
        try:
            return self._tool_handlers[name](arguments)
        except Exception as e:
            return {"error": str(e)}

    def get_toolbox_schema(self) -> list[dict]:
        """Get the toolbox tool definition for the model."""
        return [
            {
                "type": "function",
                "name": "toolbox",
                "description": """Access the toolbox to find and use tools.

Actions:
- search: Find tools by describing what you want to do
- get: Get full parameter details for a tool
- use: Execute a tool with arguments

Use 'search' to find the right tool, 'get' to see its parameters, then 'use' to execute.""",
                "parameters": {
                    "type": "object",
                    "properties": {
                        "action": {
                            "type": "string",
                            "enum": ["search", "get", "use"],
                            "description": "The action to perform",
                        },
                        "query": {
                            "type": "string",
                            "description": "Search query (for 'search' action)",
                        },
                        "tool_name": {
                            "type": "string",
                            "description": "Tool name (for 'get' and 'use' actions)",
                        },
                        "arguments": {
                            "type": "object",
                            "description": "Arguments for the tool (for 'use' action)",
                        },
                    },
                    "required": ["action"],
                },
            },
        ]

File: ui/lib/services/semantic-fs.ts
// Semantic filesystem backed by Mixedbread
import Mixedbread from "@mixedbread/sdk";

export interface FileMetadata {
  path?: string;
  type?: string;
  created_at?: string;
  updated_at?: string;
  [key: string]: unknown;
}

export interface FileResult {
  path: string;
  content: string;
  metadata: FileMetadata;
}

export interface FileError {
  error: string;
}

export interface FileListItem {
  path: string;
  metadata: FileMetadata;
}

export interface SearchResult {
  path: string;
  content: string;
  score: number;
  metadata: FileMetadata;
}

export class SemanticFS {
  private client: Mixedbread;
  private storeName: string;
  private initialized = false;

  constructor(apiKey: string, storeName = "bernd") {
    this.client = new Mixedbread({ apiKey });
    this.storeName = storeName;
  }

  private async ensureStore(): Promise<void> {
    if (this.initialized) return;

    const stores = await this.client.stores.list({ q: this.storeName });
    if (!stores.data || stores.data.length === 0) {
      await this.client.stores.create({
        name: this.storeName,
        description: "Semantic filesystem for Bernd",
        config: {
          contextualization: {
            with_metadata: [
              "type",
              "created_at",
              "updated_at",
              "priority",
              "status",
              "due_date",
              "tags",
            ],
          },
        },
      });
    }
    this.initialized = true;
  }

  private pathToId(path: string): string {
    let p = path.trim();
    if (p.startsWith("/")) {
      p = p.slice(1);
    }
    return p.replace(/\//g, "__");
  }

  private idToPath(fileId: string): string {
    return `/${fileId.replace(/__/g, "/")}`;
  }

  async write(
    path: string,
    content: string,
    metadata?: Record<string, unknown>,
  ): Promise<{ status: string; path: string }> {
    await this.ensureStore();
    const fileId = this.pathToId(path);
    const now = new Date().toISOString();

    // Preserve created_at if file exists
    let createdAt = now;
    try {
      const existing = await this.client.stores.files.retrieve(fileId, {
        store_identifier: this.storeName,
      });
      createdAt =
        ((existing.metadata as Record<string, unknown> | undefined)
          ?.created_at as string) ?? now;
      await this.client.stores.files.delete(fileId, {
        store_identifier: this.storeName,
      });
    } catch {
      // File doesn't exist, use current time
    }

    const filename = path.split("/").pop() ?? "file";
    const mimeType = filename.endsWith(".json")
      ? "text/plain"
      : "text/markdown";

    // Ensure content is not empty - Mixedbread API rejects 0-byte files
    const fileContent = content || " ";
    const blob = new Blob([fileContent], { type: mimeType });
    const file = new File([blob], filename, { type: mimeType });

    await this.client.stores.files.upload(this.storeName, file, {
      external_id: fileId,
      overwrite: true,
      metadata: {
        ...(metadata ?? {}),
        filename,
        size: blob.size,
        path,
        created_at: createdAt,
        updated_at: now,
      },
    });

    // Store metadata separately via update if supported
    // For now, encode metadata in the file content or use a different approach

    return { status: "ok", path };
  }

  async read(path: string): Promise<FileResult | FileError> {
    await this.ensureStore();
    const fileId = this.pathToId(path);

    try {
      const resp = await this.client.stores.files.retrieve(fileId, {
        store_identifier: this.storeName,
      });

      const contentResp = await this.client.files.content(resp.id);
      const content = await contentResp.text();

      return {
        path,
        content,
        metadata: (resp.metadata as FileMetadata) ?? {},
      };
    } catch (e) {
      console.error(`[SemanticFS] Error reading ${path}:`, e);
      return { error: `Not found: ${path}` };
    }
  }

  async delete(
    path: string,
  ): Promise<{ status: string; path: string } | FileError> {
    await this.ensureStore();
    const fileId = this.pathToId(path);

    try {
      await this.client.stores.files.delete(fileId, {
        store_identifier: this.storeName,
      });
      return { status: "deleted", path };
    } catch {
      return { error: `Not found: ${path}` };
    }
  }

  async list(prefix = "/", limit = 100): Promise<FileListItem[]> {
    await this.ensureStore();
    const prefixId = prefix !== "/" ? this.pathToId(prefix) : "";
    const files: FileListItem[] = [];
    let cursor: string | undefined;

    while (files.length < limit) {
      const resp = await this.client.stores.files.list(this.storeName, {
        limit: Math.min(100, limit - files.length),
        after: cursor,
      });

      for (const item of resp.data ?? []) {
        const externalId = item.external_id ?? "";
        if (prefix === "/" || externalId.startsWith(prefixId)) {
          files.push({
            path: this.idToPath(externalId),
            metadata: (item.metadata as FileMetadata) ?? {},
          });
        }
      }

      if (!resp.pagination?.has_more) break;
      cursor = resp.pagination?.last_cursor ?? undefined;
    }

    return files;
  }

  async search(
    query: string,
    prefix = "/",
    topK = 10,
  ): Promise<SearchResult[]> {
    await this.ensureStore();

    // Build metadata filter for prefix
    const filters =
      prefix !== "/"
        ? {
            key: "path",
            operator: "starts_with" as const,
            value: prefix,
          }
        : undefined;

    try {
      const resp = await this.client.stores.search({
        store_identifiers: [this.storeName],
        query,
        top_k: topK,
        filters,
      });

      return (resp.data ?? []).map((item) => {
        const metadata = (item.metadata as FileMetadata) ?? {};
        const path = metadata.path ?? `/${item.filename ?? ""}`;

        // Handle different chunk types
        let content = "";
        if ("text" in item && typeof item.text === "string") {
          content = item.text;
        }

        return {
          path,
          content,
          score: item.score ?? 0,
          metadata,
        };
      });
    } catch (e) {
      console.error("[SemanticFS] Search error:", e);
      return [];
    }
  }

  async clearPrefix(
    prefix: string,
  ): Promise<{ status: string; prefix: string; deleted: number }> {
    const files = await this.list(prefix, 1000);
    let deleted = 0;

    for (const f of files) {
      await this.delete(f.path);
      deleted++;
    }

    return { status: "cleared", prefix, deleted };
  }

  async writeBinary(
    path: string,
    data: ArrayBuffer,
    mimeType: string,
    metadata?: Record<string, unknown>,
  ): Promise<{ status: string; path: string }> {
    await this.ensureStore();
    const fileId = this.pathToId(path);
    const now = new Date().toISOString();

    // Delete existing file if present
    try {
      await this.client.stores.files.delete(fileId, {
        store_identifier: this.storeName,
      });
    } catch {
      // File doesn't exist
    }

    const filename = path.split("/").pop() ?? "file";
    const blob = new Blob([data], { type: mimeType });
    const file = new File([blob], filename, { type: mimeType });

    await this.client.stores.files.upload(this.storeName, file, {
      external_id: fileId,
      overwrite: true,
      metadata: {
        ...(metadata ?? {}),
        path,
        created_at: now,
        type: "binary",
      },
    });

    return { status: "ok", path };
  }

  async readBinary(
    path: string,
  ): Promise<
    { path: string; data: ArrayBuffer; metadata: FileMetadata } | FileError
  > {
    await this.ensureStore();
    const fileId = this.pathToId(path);

    try {
      const resp = await this.client.stores.files.retrieve(fileId, {
        store_identifier: this.storeName,
      });

      const contentResp = await this.client.files.content(resp.id);
      const data = await contentResp.arrayBuffer();

      return {
        path,
        data,
        metadata: (resp.metadata as FileMetadata) ?? {},
      };
    } catch (e) {
      console.error(`[SemanticFS] Error reading binary ${path}:`, e);
      return { error: `Not found: ${path}` };
    }
  }
}
